File: HardwareID.cs
public class HardwareID
{
    public string LastID { get; private set; } = string.Empty;

    public string Generate()
    {
        var wmiData = new[]
        {
            GetProperties(WMI_CLASSES.MOTHERBOARD),
            GetProperties(WMI_CLASSES.CDROM),
            GetProperties(WMI_CLASSES.BIOS),
            GetProperties(WMI_CLASSES.GPU),
            GetProperties(WMI_CLASSES.CPU),
            GetProperties(WMI_CLASSES.HDD),
            GetProperties(WMI_CLASSES.ACCOUNT)
        };
        return LastID = GetHash(data: string.Join(string.Empty, wmiData));
    }

    private static string GetProperties(string[] wmiData)
    {
        var property = new StringBuilder();
        string query = CreateQuery(wmiData);
        using (var moSearcher = new ManagementObjectSearcher("root\\CIMV2", query))
        using (ManagementObjectCollection moCollection = moSearcher.Get())
        {
            foreach (ManagementBaseObject mbObject in moCollection)
                using (mbObject)
                {
                    for (int i = 1; i < wmiData.Length; i++)
                        property.Append(value: mbObject[propertyName: wmiData[i]]);
                }
        }

        return property.ToString();
    }

    private static string CreateQuery(string[] wmiData)
    {
        if (wmiData == null)
            throw new ArgumentNullException(paramName: nameof(wmiData));
        var query = new StringBuilder();
        string wmiClass = string.Empty;
        query.Append("SELECT ");
        for (int i = 0; i < wmiData.Length; i++)
            if (i == 0)
                wmiClass = wmiData[i];
            else
                query.Append(value: i < wmiData.Length - 1 ? $"{wmiData[i]}, " : $"{wmiData[i]} ");
        query.Append(value: $"FROM {wmiClass}");
        return query.ToString();
    }

    private static string GetHash(string data)
    {
        using (var sha = new SHA1CryptoServiceProvider())
        {
            var hash = sha.ComputeHash(buffer: Encoding.Default.GetBytes(data));
            return GetHexString(hash);
        }
    }

    private static string GetHexString(IList<byte> bt)
    {
        var sBuilder = new StringBuilder();
        for (int i = 0; i < bt.Count; i++)
        {
            byte b = bt[i];
            byte n = b;
            int n1 = n & 15;
            int n2 = (n >> 4) & 15;
            sBuilder.Append(value: n2 > 9
                ? ((char) (n2 - 10 + 'A')).ToString(CultureInfo.InvariantCulture)
                : n2.ToString(CultureInfo.InvariantCulture));
            sBuilder.Append(value: n1 > 9
                ? ((char) (n1 - 10 + 'A')).ToString(CultureInfo.InvariantCulture)
                : n1.ToString(CultureInfo.InvariantCulture));
            if (i + 1 != bt.Count && (i + 1) % 2 == 0)
                sBuilder.Append("-");
        }

        return sBuilder.ToString();
    }

    private static class WMI_CLASSES
    {
        public static readonly string[] MOTHERBOARD = {"Win32_BaseBoard", "Name", "Manufacturer", "Version"};
        public static readonly string[] GPU = {"Win32_VideoController", "Name", "DeviceID", "DriverVersion"};
        public static readonly string[] CDROM = {"Win32_CDROMDrive", "Name", "Manufacturer", "DeviceID"};
        public static readonly string[] CPU = {"Win32_Processor", "Name", "Manufacturer", "ProcessorId"};
        public static readonly string[] HDD = {"Win32_DiskDrive", "Name", "Manufacturer", "Model"};
        public static readonly string[] BIOS = {"Win32_BIOS", "Name", "Manufacturer", "SerialNumber"};
        public static readonly string[] ACCOUNT = {"Win32_UserAccount", "Name"};
    }
}

File: HardwareID.vb
Public Class HardwareID
    Protected Class WMI_CLASSES
        Public Shared ReadOnly MOTHERBOARD As String() = {"Win32_BaseBoard", "Name", "Manufacturer", "Version"}
        Public Shared ReadOnly GPU As String() = {"Win32_VideoController", "Name", "DeviceID", "DriverVersion"}
        Public Shared ReadOnly CDROM As String() = {"Win32_CDROMDrive", "Name", "Manufacturer", "DeviceID"}
        Public Shared ReadOnly CPU As String() = {"Win32_Processor", "Name", "Manufacturer", "ProcessorId"}
        Public Shared ReadOnly HDD As String() = {"Win32_DiskDrive", "Name", "Manufacturer", "Model"}
        Public Shared ReadOnly BIOS As String() = {"Win32_BIOS", "Name", "Manufacturer", "Version"}
    End Class

    Public Property LastID As String

    Sub New() : End Sub

    Public Function Generate() As String
        Dim wmiData As String() = {
            GetProperties(WMI_CLASSES.MOTHERBOARD),
            GetProperties(WMI_CLASSES.CDROM),
            GetProperties(WMI_CLASSES.BIOS),
            GetProperties(WMI_CLASSES.GPU),
            GetProperties(WMI_CLASSES.CPU),
            GetProperties(WMI_CLASSES.HDD)
        }
        LastID = GetHash(String.Join(String.Empty, wmiData))
        Return LastID
    End Function

    Private Function GetProperties(ByVal wmiData As String()) As String
        Dim properties As New StringBuilder()
        Dim query As String = GenerateQuery(wmiData)
        Using moSearcher As New ManagementObjectSearcher("root\CIMV2", query)
            Using moCollection As ManagementObjectCollection = moSearcher.Get()
                For Each mbObject As ManagementObject In moCollection
                    Using mbObject
                        For i As Integer = 1 To wmiData.Length - 1
                            properties.Append(mbObject(wmiData(i)))
                        Next
                    End Using
                Next
            End Using
        End Using
        Return properties.ToString()
    End Function

    Private Function GenerateQuery(ByVal wmiData As String()) As String
        Dim query As New StringBuilder()
        Dim wmiClass As String = String.Empty
        query.Append("SELECT ")
        For i As Integer = 0 To wmiData.Length - 1
            If (i = 0) Then
                wmiClass = wmiData(i)
            Else
                query.Append(If(i < wmiData.Length - 1, $"{wmiData(i)}, ", $"{wmiData(i)} "))
            End If
        Next
        query.Append($"FROM {wmiClass}")
        Return query.ToString()
    End Function

    Private Function GetHash(ByVal data As String) As String
        Using SHA As New SHA1CryptoServiceProvider()
            Dim hash As Byte() = SHA.ComputeHash(Encoding.Default.GetBytes(data))
            Return GetHexString(hash)
        End Using
    End Function

    Private Function GetHexString(ByVal bt As Byte()) As String
        Dim sBuilder = New StringBuilder()
        For i As Integer = 0 To bt.Count - 1
            Dim b = bt(i)
            Dim n = b
            Dim n1 = n And 15
            Dim n2 = (n >> 4) And 15
            If n2 > 9 Then
                sBuilder.Append(ChrW((n2 - 10 + AscW("A"))).ToString(CultureInfo.InvariantCulture))
            Else
                sBuilder.Append(n2.ToString(CultureInfo.InvariantCulture))
            End If
            If n1 > 9 Then
                sBuilder.Append(ChrW((n1 - 10 + AscW("A"c))).ToString(CultureInfo.InvariantCulture))
            Else
                sBuilder.Append(n1.ToString(CultureInfo.InvariantCulture))
            End If
            If (i + 1) <> bt.Count AndAlso (i + 1) Mod 2 = 0 Then
                sBuilder.Append("-")
            End If
        Next
        Return sBuilder.ToString()
    End Function
End Class